R/derive_param_tte.R

```diff
@@ -613,17 +642,50 @@ filter_date_sources <- function(sources,
       var = !!source_date_var,
       dataset_name = sources[[i]]$dataset_name
     )
-    data[[i]] <- source_dataset %>%
-      filter_if(sources[[i]]$filter) %>%
-      filter_extreme(
-        order = exprs(!!source_date_var),
-        by_vars = expr_c(subject_keys, by_vars),
-        mode = mode,
-        check_type = "none"
-      )
-
+    # wrap filter_extreme in tryCatch to catch duplicate records and create a message
+    data[[i]] <- rlang::try_fetch(
+      {
+        source_dataset %>%
+          filter_if(sources[[i]]$filter) %>%
+          arrange(!!!sources[[i]]$order) %>% # Ensure order is applied
+          filter_extreme(
+            order = expr_c(exprs(!!source_date_var), sources[[i]]$order),
+            by_vars = expr_c(subject_keys, by_vars),
+            mode = mode,
+            check_type = check_type
+          )
+      },
+      warning = function(cnd) {
+        # Handle warnings
+        if (grepl("duplicate records", conditionMessage(cnd))) {
+          cli::cli_warn(c(
+            "Dataset '{.val {sources[[i]]$dataset_name}}' contains duplicate records.",
+            "i Duplicates were identified based on variables:
+               {.val {paste(c(subject_keys, by_vars, source_date_var), collapse = ', ')}}."
+          ))
+        }
+        source_dataset %>%
+          filter_if(sources[[i]]$filter) %>%
+          arrange(!!!sources[[i]]$order) # Return filtered dataset even if a warning occurred
+      },
+      error = function(err) {
+        cli::cli_abort(c(
+          "Duplicate records detected during processing.",
+          "x Duplicate records were found in dataset {.val {sources[[i]]$dataset_name}}.",
+          "i The duplicates were identified based on the following variables:
+            {.val {paste(c(subject_keys, by_vars, source_date_var), collapse = ', ')}}.",
+          "i Consider reviewing your `by_vars` or `order` argument to ensure uniqueness."
+        ))
+      },
+      message = function(msg) {
+        cli::cli_inform(c(
+          "Processing dataset '{.val {sources[[i]]$dataset_name}}'...",
+          "i Filter and order criteria: {.val {paste(c(subject_keys, by_vars,
+          sources[[i]]$order), collapse = ', ')}}."
+        ))
+      }
```

Comment:
Once https://github.com/pharmaverse/admiral/pull/2592 is merged this could be simplified:
```
      duplicate_records = function(cnd) {
        cnd_funs <- list(message = cli_inform, warning = cli_warn, error = cli_abort)
        cnd_funs[[check_type]](
          paste(
            "Dataset {.val {sources[[i]]$dataset_name}} contains duplicate records with respect to",
            "{.var {cnd$by_vars}}"
          ),
          class = class(cnd))
        cnd_muffle(cnd)
        zap()
      }
```